veros/core/external/solvers/petsc_.py

```diff
@@ -41,7 +74,34 @@ def __init__(self, state):
             self._da.setVecType("cuda")
             self._da.setMatType("aijcusparse")
 
-        self._matrix, self._boundary_fac = self._assemble_poisson_matrix(state)
+        diags, offsets = assemble_poisson_matrix(state)
+        diags = onp.asarray(diags, dtype=onp.float64)
+        diags = diags[:, 2:-2, 2:-2]
+
+        row = PETSc.Mat.Stencil()
+        col = PETSc.Mat.Stencil()
+
+        (i0, i1), (j0, j1) = self._da.getRanges()
+        matrix = self._da.getMatrix()
+
+        for j in range(j0, j1):
+            for i in range(i0, i1):
+                iloc, jloc = i % (settings.nx // rs.num_proc[0]), j % (settings.ny // rs.num_proc[1])
+                row.index = (i, j)
+                for diag, offset in zip(diags, offsets):
+                    io, jo = (i + offset[0], j + offset[1])
+                    col.index = (io, jo)
+                    matrix.setValueStencil(row, col, diag[iloc, jloc])
+
+        matrix.assemble()
```

Comment:
Move this to `self._assemble_poisson_matrix` (like SciPy solver).

Reply:
Done, didn't push for a couple of days.